crates/runtime/src/traphandlers.rs

```diff
@@ -397,3 +429,120 @@ mod tls {
+            let guard_size = page_size;
+
+            let mut me = Self {
+                guard_size,
+                sigaltstack_alloc_size: guard_size + STACK_SIZE,
+                new_stack: libc::stack_t {
+                    ss_sp: null_mut(),
+                    ss_flags: 0,
+                    ss_size: 0,
+                },
+            };
+
+            // Allocate memory.
+            let ptr = libc::mmap(
+                null_mut(),
+                me.sigaltstack_alloc_size,
+                libc::PROT_NONE,
+                libc::MAP_PRIVATE | libc::MAP_ANON,
+                -1,
+                0,
+            );
+            if ptr == libc::MAP_FAILED {
+                super::raise_oom_trap()
+            }
+
+            // Prepare the stack, register it, and sanity check the old stack.
+            let stack_ptr = (ptr as usize + me.guard_size) as *mut libc::c_void;
+            me.new_stack = libc::stack_t {
+                ss_sp: stack_ptr,
+                ss_flags: 0,
+                ss_size: STACK_SIZE,
+            };
+            let r = libc::mprotect(stack_ptr, STACK_SIZE, libc::PROT_READ | libc::PROT_WRITE);
+            assert_eq!(r, 0, "mprotect to configure memory for sigaltstack failed");
+
+            // Register the new alternate stack (and record the old one, so we can sanity-check it).
+            let mut old_stack = libc::stack_t {
+                ss_sp: null_mut(),
+                ss_flags: 0,
+                ss_size: 0,
+            };
+            let r = libc::sigaltstack(&mut me.new_stack, &mut old_stack);
+            assert_eq!(r, 0, "registering new sigaltstack failed");
+
+            // Sanity-check the old stack.
+            assert_eq!(
+                old_stack.ss_flags, 0,
+                "old sigaltstack had unexpected flags set"
+            );
+            assert!(
+                old_stack.ss_size <= STACK_SIZE,
+                "old sigaltstack had a larger stack"
+            );
+
+            me
+        }
+    }
+
+    impl Drop for Tls {
+        fn drop(&mut self) {
+            unsafe {
+                // We expect Rust's libstd to have unregistered our sigaltstack by this point.
+                // Confirm that it has.
+                let mut old_stack = libc::stack_t {
+                    ss_sp: null_mut(),
+                    ss_flags: 0,
+                    ss_size: 0,
+                };
+                let r = libc::sigaltstack(null_mut(), &mut old_stack);
+                assert_eq!(r, 0, "querying for our sigaltstack failed");
+                assert_ne!(
+                    old_stack.ss_sp, self.new_stack.ss_sp,
+                    "Rust's runtime didn't unregister our sigaltstack"
+                );
+
+                // Deallocate the stack memory.
+                let alloc_ptr =
+                    (self.new_stack.ss_sp as usize - self.guard_size) as *mut libc::c_void;
```

Comment:
Since `Tls` is constructed with a null `ss_sp` before we call `mmap`, I think this'll want to check for null and bail out because `mmap` would have failed.